frame/uniques/src/tests.rs

```diff
@@ -0,0 +1,469 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2019-2021 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Tests for Uniques pallet.
+
+use super::*;
+use crate::mock::*;
+use frame_support::{assert_ok, assert_noop, traits::Currency};
+use pallet_balances::Error as BalancesError;
+
+fn assets() -> Vec<(u64, u32, u32)> {
+	let mut r: Vec<_> = Account::<Test>::iter().map(|x| x.0).collect();
+	r.sort();
+	let mut s: Vec<_> = Asset::<Test>::iter().map(|x| (x.2.owner, x.0, x.1)).collect();
+	s.sort();
+	assert_eq!(r, s);
+	for class in Asset::<Test>::iter()
+		.map(|x| x.0)
+		.scan(None, |s, item| if s.map_or(false, |last| last == item) {
+				*s = Some(item);
+				Some(None)
+			} else {
+				Some(Some(item))
+			}
+		).filter_map(|item| item)
+	{
+		let details = Class::<Test>::get(class).unwrap();
+		let instances = Asset::<Test>::iter_prefix(class).count() as u32;
+		assert_eq!(details.instances, instances);
+	}
+	r
+}
+
+#[test]
+fn basic_setup_works() {
+	new_test_ext().execute_with(|| {
+		assert_eq!(assets(), vec![]);
+	});
+}
+
+#[test]
+fn basic_minting_should_work() {
+	new_test_ext().execute_with(|| {
+		assert_ok!(Uniques::force_create(Origin::root(), 0, 1, true));
+		assert_ok!(Uniques::mint(Origin::signed(1), 0, 42, 1));
+		assert_eq!(assets(), vec![(1, 0, 42)]);
+
+		assert_ok!(Uniques::force_create(Origin::root(), 1, 2, true));
+		assert_ok!(Uniques::mint(Origin::signed(2), 1, 69, 1));
+		assert_eq!(assets(), vec![(1, 0, 42), (1, 1, 69)]);
+	});
+}
+
+#[test]
+fn lifecycle_should_work() {
+	new_test_ext().execute_with(|| {
+		Balances::make_free_balance_be(&1, 100);
+		assert_ok!(Uniques::create(Origin::signed(1), 0, 1));
+		assert_eq!(Balances::reserved_balance(&1), 2);
+
+		assert_ok!(Uniques::set_class_metadata(Origin::signed(1), 0, vec![0], vec![0], false));
```

Comment:
In case you refactor this code for using `BoundedVec`, you can use this macro to facilitate the refactor:

``` rust
macro_rules! bounded_vec {
    () => (
        <BoundedVec::<u8, <mock::Test as Config>::StringLimit>>::default()
    );
    ($elem:expr; $n:expr) => (
        unsafe {
            <BoundedVec<u8, <Test as Config>::StringLimit>>::force_from(vec![$elem; $n], None)
        }
    );
    ($($x:expr),+ $(,)?) => (
        unsafe {
            <BoundedVec::<u8, <Test as Config>::StringLimit>>::force_from(vec![$($x),+], None)
        }
    );
}
```

```suggestion
		assert_ok!(Uniques::set_class_metadata(Origin::signed(1), 0, bounded_vec![0], bounded_vec![0], false));
```